File: basics/02-linkedlist/0202-86.cpp
/*
 * leetcode 86
 *
 * must break original linked list
 * or there will be a circle
 *
 */

struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* dummy1 = new ListNode(-1);
        ListNode* dummy2 = new ListNode(-1);
        ListNode* p = head;
        ListNode* p1 = dummy1;
        ListNode* p2 = dummy2;
        while(p != nullptr) {
            if (p->val < x) {
                p1->next = p;
                p1 = p1->next;
            } else {
                p2->next = p;
                p2 = p2->next;
            }
            // break p
            ListNode* t = p->next;
            p->next = nullptr;
            p = t;
        }
        p1->next = dummy2->next;
        return dummy1->next;    
    }
};

File: basics/02-linkedlist/0202-86.go
/*
 *
 */

type ListNode struct {
    Val int
    Next *ListNode
}

func partition(head *ListNode, x int) *ListNode {
    new_head := &ListNode{0,head}
    pos := new_head
    prev := new_head
    for head != nil{
        if head.Val >= x{
            prev = prev.Next
            head = head.Next
        }else if pos == prev{
            pos = pos.Next
            prev = prev.Next
            head = head.Next
        }else{
            head_next := head.Next
            prev.Next = head_next
            pos_next := pos.Next
            pos.Next = head
            head.Next = pos_next
            head = head_next
            pos = pos.Next
        }
    }
    return new_head.Next    
}

